lib/services/badges_S/badge_service.dart

```diff
@@ -0,0 +1,487 @@
+          DateTime latestDate =
+              (data["loyalty_badge"]["latest_date"] as Timestamp).toDate();
+
+          ///get the latest date from the document
+          int count =
+              data["loyalty_badge"]["count"]; //get the count from the document
+
+          // check if the function has already run today
+          bool isSameDay = latestDate.day == currentDate.day &&
+              latestDate.month == currentDate.month &&
+              latestDate.year == currentDate.year;
+
+          if (isSameDay) {
+            //if it is the same day, then don't run again
+            return;
+          }
+
+          bool isYesterday = latestDate.day == currentDate.day - 1 &&
+              latestDate.month == currentDate.month &&
+              latestDate.year == currentDate.year;
+
+          if (isYesterday) {
+            //if the latest date was yesterday then increment the count
+            count = count + 1;
+          } else {
+            count = 0;
+          }
+
+          if (count >= 10) {
+            await _firestore.collection("badges").doc(currentUser.uid).update({
+              //if the count is >= 10 then update the document with the unlocked status
+              "loyalty_badge": {
+                "count": count,
+                "latest_date": currentDate,
+                "date_unlocked": currentDate,
+                "unlocked": true,
+              }
+            });
+          } else {
+            await _firestore.collection("badges").doc(currentUser.uid).update({
+              //update the count and latest date in the document
+              "loyalty_badge": {
+                "count": count,
+                "latest_date": currentDate,
+                "date_unlocked": null,
+                "unlocked": false,
+              }
+            });
+          }
+        }
+      }
+    } catch (e) {
+      throw Exception("Failed to unlock loyalty badge: $e");
+    }
+  }
+
+  void unlockCollectorBadge(bool add) async {
+    try {
+      final currentUser = _auth.currentUser;
+
+      if (currentUser != null) {
+        DocumentSnapshot<Map<String, dynamic>> doc =
+            await _firestore.collection("badges").doc(currentUser.uid).get();
+        Map<String, dynamic>? data = doc.data();
+
+        if (data != null) {
+          int count = data["collector_badge"]
+              ["count"]; //get the count from the document//CHANGE
+          bool unlocked = data["collector_badge"]["unlocked"];
+
+          if (unlocked == false) {
+            // only update if its false
+            if (add) {
+              count++;
+            } else {
+              count--;
+            }
+
+            if (count == 10) {
+              await _firestore
+                  .collection("badges")
+                  .doc(currentUser.uid)
+                  .update({
+                //if the count is == 10 then update the document with the unlocked status
+                "collector_badge": {
+                  //CHANGE
```

Comment:
Not too sure what this comment means, just wanted to point it out incase you left this comment as a reminder to chane something 😊